src/frontend/xgboost_json.cc

```diff
@@ -42,27 +42,62 @@ namespace frontend {
 DMLC_REGISTRY_FILE_TAG(xgboost_json);
 
 std::unique_ptr<treelite::Model> LoadXGBoostJSONModel(const char* filename) {
-  char readBuffer[65536];
+  char read_buffer[65536];
 
 #ifdef _WIN32
-  FILE * fp = fopen(filename, "rb");
+  FILE* fp = std::fopen(filename, "rb");
 #else
-  FILE * fp = fopen(filename, "r");
+  FILE* fp = std::fopen(filename, "r");
 #endif
 
-  auto input_stream = std::make_unique<rapidjson::FileReadStream> (
-      fp,
-      readBuffer,
-      sizeof(readBuffer));
-  auto parsed_model = ParseStream(std::move(input_stream));
-  fclose(fp);
+  auto input_stream = std::make_unique<rapidjson::FileReadStream>(
+      fp, read_buffer, sizeof(read_buffer));
+  auto error_handler = [fp](size_t offset) -> std::string {
+    size_t cur = (offset >= 50 ? (offset - 50) : 0);
+    std::fseek(fp, cur, SEEK_SET);
+    int c;
+    std::ostringstream oss, oss2;
+    for (int i = 0; i < 100; ++i) {
+      c = std::fgetc(fp);
+      if (c == EOF) {
+        break;
+      }
+      oss << static_cast<char>(c);
+      if (cur == offset) {
+        oss2 << "^";
+      } else {
+        oss2 << "~";
+      }
+      ++cur;
+    }
+    std::fclose(fp);
+    return oss.str() + "\n" + oss2.str();
+  };
+  auto parsed_model = ParseStream(std::move(input_stream), error_handler);
+  std::fclose(fp);
   return parsed_model;
 }
 
 std::unique_ptr<treelite::Model> LoadXGBoostJSONModelString(const char* json_str, size_t length) {
-  auto input_stream = std::make_unique<rapidjson::MemoryStream>(
-      json_str, length);
-  return ParseStream(std::move(input_stream));
+  auto input_stream = std::make_unique<rapidjson::MemoryStream>(json_str, length);
+  auto error_handler = [json_str](size_t offset) -> std::string {
+    size_t cur = (offset >= 50 ? (offset - 50) : 0);
+    std::ostringstream oss, oss2;
+    for (int i = 0; i < 100; ++i) {
+      if (!json_str[cur]) {
+        break;
+      }
+      oss << json_str[cur];
```

Comment:
Any particular reason to break this up into two streams as opposed to just doing `oss << "\n"` and then continuing with everything you build into oss2?

Reply:
Wait, nevermind! Ignore this. I was looking at it wrong.

Reply:
`oss2` is for creating the second line `~~~~~~~~^~~~~~~~`, which is super useful to locate the exact location of the error.